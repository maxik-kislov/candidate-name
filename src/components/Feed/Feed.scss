.feed {
  margin: 0;
  padding: 8px;
  background-color: #f3ce56;
  box-sizing: border-box;
  max-width: 680px;

  &__title {
    margin: 0 0 15px;
    font-size: 24px;
    font-weight: 500;
    line-height: 2rem;
  }
}

@media (min-width: 30em) {
  .feed {
    padding: 24px;
    padding-bottom: 1px;
    margin-bottom: 64px;
    letter-spacing: 0;
  }
}
