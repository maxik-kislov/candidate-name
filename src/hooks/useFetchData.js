import { useEffect, useState } from 'react';

const API_BASE_URL =
  'https://magiclab-twitter-interview.herokuapp.com/andreupifarre';

const useFetchData = (trigger, count = 20, resetFromId = null) => {
  const [results, setResults] = useState([]);
  const [afterId, setAfterId] = useState(0);

  // stop fething on scrolling
  // start fetching when on top
  // bottom request for tweetsbefore that id

  useEffect(() => {
    const fetchData = async () => {
      try {
        let url = '';
        if (afterId > 0) {
          url = `${API_BASE_URL}/api?count=${count}&afterId=${afterId}`;
        } else {
          url = `${API_BASE_URL}/api?count=${count}`;
        }
        const response = await fetch(url);

        // if (afterId >= resetFromId) {
        //   await fetch(`${API_BASE_URL}/reset`);
        //   setAfterId(0);
        //   setResults([]);
        //   return;
        // }

        if (response.ok) {
          const data = await response.json();

          if (data.length > 0) {
            setAfterId(data[0].id);
          }
          setResults([...data, ...results]);
        }
      } catch (err) {
        console.error(err);
      }
    };

    fetchData();
    // eslint-disable-next-line
  }, [trigger, count]);

  return results;
};

export default useFetchData;
