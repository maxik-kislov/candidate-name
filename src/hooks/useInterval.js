import { useState, useEffect } from 'react';

const useInterval = (delay, stop) => {
  const [count, setCount] = useState(0);

  useEffect(() => {
    const timer = setInterval(() => {
      if (!stop) setCount(count + 1);
    }, delay);
    return () => clearInterval(timer);
  }, [delay, count, stop]);

  return count;
};

export default useInterval;
