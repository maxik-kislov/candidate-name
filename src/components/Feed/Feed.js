import React, { useState, useEffect, useRef } from 'react';
import List from '../List/List';
import useFetchData from '../../hooks/useFetchData';
import useInterval from '../../hooks/useInterval';
import './Feed.scss';

const Feed = () => {
  const [stop, setStop] = useState(false);
  const refFeed = useRef(null);
  const interval = useInterval(2000, stop);
  const tweets = useFetchData(interval);

  useEffect(() => {
    window.addEventListener('scroll', e => {
      // console.log(window.scrollY);
      setStop(window.scrollY > 0);
      console.log(
        window.scrollY,
        refFeed.current.clientHeight,
        window.scrollY - refFeed.current.clientHeight,
      );
    });
  }, []);

  return (
    <div className="feed" ref={refFeed}>
      <h2 className="feed__title">Twitter-like feed</h2>
      <List tweets={tweets} />
    </div>
  );
};

export default Feed;
