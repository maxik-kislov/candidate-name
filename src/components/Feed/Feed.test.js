import setupTests from '../../setupTests';
import React from 'react';
import { shallow } from 'enzyme';

import Feed from './Feed';
import useInterval from '../../hooks/useInterval';
import useFetchData from '../../hooks/useFetchData';

jest.mock('../../hooks/useInterval');
jest.mock('../../hooks/useFetchData');

useInterval.mockReturnValue('');
useFetchData.mockReturnValue([]);

const render = () => shallow(<Feed />);

afterEach(() => {
  jest.clearAllMocks();
});

describe('Feed', () => {
  it('should contain an element with a class .feed', () => {
    expect(
      render()
        .find('.feed')
        .exists(),
    ).toBeTruthy();
  });

  it('should contain an H2 element with a class .feed__title', () => {
    const element = render().find('.feed__title');
    expect(element.exists()).toBeTruthy();
    expect(element.name()).toEqual('h2');
  });

  it('should contain a List component', () => {
    const element = render().find('List');
    expect(element.exists()).toBeTruthy();
    expect(element.prop('tweets')).toBeDefined();
  });

  it('should pass the correct object to List tweets property', () => {
    const tweets = [
      {
        image: 'https://s3.amazonaws.com/uifaces/faces/twitter/pixage/128.jpg',
        text:
          '10001. "It is straight enough. I should like to know who sold you the\ngeese which you supplied to the Alpha."',
        username: 'Ferne.Dickinson1',
      },
      {
        image:
          'https://s3.amazonaws.com/uifaces/faces/twitter/kazaky999/128.jpg',
        text:
          '10000. Godfrey\nNorton, as our client is to its coming to the eyes of his\nprincess.',
        username: 'Jena_Heller',
      },
    ];

    useInterval.mockReturnValue(0);
    useFetchData.mockReturnValue(tweets);

    expect(
      render()
        .find('List')
        .prop('tweets'),
    ).toEqual(tweets);
  });

  it('should render correctly', () => {
    expect(render().getElements()).toMatchSnapshot();
  });
});
