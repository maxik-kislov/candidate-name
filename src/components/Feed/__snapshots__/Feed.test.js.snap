// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed should render correctly 1`] = `
Array [
  <div
    className="feed"
  >
    <h2
      className="feed__title"
    >
      Twitter-like feed
    </h2>
    <List
      tweets={
        Array [
          Object {
            "image": "https://s3.amazonaws.com/uifaces/faces/twitter/pixage/128.jpg",
            "text": "10001. \\"It is straight enough. I should like to know who sold you the
geese which you supplied to the Alpha.\\"",
            "username": "Ferne.Dickinson1",
          },
          Object {
            "image": "https://s3.amazonaws.com/uifaces/faces/twitter/kazaky999/128.jpg",
            "text": "10000. Godfrey
Norton, as our client is to its coming to the eyes of his
princess.",
            "username": "Jena_Heller",
          },
        ]
      }
    />
  </div>,
]
`;
